packages/editor/src/store/selectors.js

```diff
@@ -138,7 +89,7 @@ export function isEditedPostDirty( state ) {
 	// Edits should contain only fields which differ from the saved post (reset
 	// at initial load and save complete). Thus, a non-empty edits state can be
 	// inferred to contain unsaved values.
-	if ( Object.keys( state.editor.present.edits ).length > 0 ) {
```

Comment:
Do we still have undo history for edited post properties?

**Edit:** No, it's entirely broken. Try editing the title. There should be undo history.

Reply:
I actually thought I'd map undo/redo to the classic editor's undo/redo and leave it only for the blocks. I think it's still possible to include the edit post if I move the undo/redo handling in the Editor store, but I expect more complexity introduced.

Reply:
>if I move the undo/redo handling in the Editor store, but I expect more complexity introduced.

I'm actually wondering if it might make sense for an idea of a separate "history" module which both `editor` and `block-editor` could opt in to, which could also develop to include ideas around history manipulation (#12724, #8119) and a rearchitecture of history to a sequence of actions (https://github.com/WordPress/gutenberg/issues/11782#issuecomment-448344911).

Reply:
Seems appealing, any ideas how the API might look like?

Reply:
Brain dump:

In my mind, stores to could opt in to it, perhaps in a similar manner to store persistence with a combination of a data plugin and `registerStore` property.

```js
wp.data.use( history() );

wp.data.registerStore( 'core/editor', {
	history: /* ... */
} );
```

I think it'd either configure to specify a path to monitor, similar to how we currently observe for changes the `state.editor` path in `core/editor`.

```js
history: [ 'editor', 'edits' ],
```

Alternatively, it could provide a callback function which receives an action being dispatched (in some form) and returns `true` if that action is to be added to the log.

```js
history: ( action ) => action.type === 'INSERT_BLOCK',
```

Alternatively yet, if we wanted to solve some of the issues where we create too many history entries, we could make it an opt-in, where the action itself might include some meta describing that it ought to be included.

```js
const insertBlock = ( block ) => ( {
	type: 'INSERT_BLOCK',
	block,
	meta: { history: true },
} );
```

It could in some ways be similar to [Calypso's Action Log](https://github.com/Automattic/wp-calypso/blob/master/client/state/action-log/README.md) as far as a general structure for storage, i.e. effectively an array of actions (either the action objects, or the store/action name + arguments pairing)

```js
history: [
	{ store: 'core/editor', action: { type: 'EDIT_POST', /* ... */ } },
	{ store: 'core/block-editor', action: { type: 'INSERT_BLOCK', /* ... */ } },
]
```

To support #12724 / #8119, we'd probably need some way both to search within the history to find specific entries, and be able to manipulate those entries. This could be done through a combination of selectors / actions offered by the history module itself.

```js
const imageInsert = wp.data.select( 'core/history' ).getLastMatchingEntry( ( action ) => {
	return action.type === 'INSERT_BLOCK' && action.clientId = currentImageBlockId;
} );
const newAction = merge( {}, imageInsert.action, attributes: { url: persistedImageURL } } );
wp.data.dispatch( 'core/history' ).replaceEntry( imageInsert.entryId, newAction );
```

In writing this, it feels like perhaps the raw actions shouldn't be surfaced to a consumer of the selector. Therefore, we might consider either the idea of tracking action creator name + arguments or, alternatively, maybe the predicate could receive a pre-bound `select` function to select data _as if_ they were at that point in history.

```js
wp.data.select( 'core/history' ).getLastMatchingEntry( ( actionName, ...args ) => {
	if ( actionName !== 'insertBlock' ) {
		return false;
	}

	const [ block ] = args;
	return block.clientId === currentImageBlockId;
} );
```